src/DataGrid.tsx

```diff
@@ -154,52 +175,83 @@ function DataGrid<R, K extends keyof R, SR>({
   cellNavigationMode = CellNavigationMode.NONE,
   editorPortalTarget = document.body,
   defaultFormatter = ValueFormatter,
-  columns,
+  columns: rawColumns,
   rows,
+  rowRenderer,
+  rowGroupRenderer,
+  summaryRows,
   selectedRows,
+  onRowClick,
+  onRowExpandToggle,
+  onSelectedCellChange,
+  onSelectedCellRangeChange,
   onSelectedRowsChange,
   ...props
 }: DataGridProps<R, K, SR>, ref: React.Ref<DataGridHandle>) {
-  const [columnWidths, setColumnWidths] = useState<ReadonlyMap<string, number>>(() => new Map());
-  const [scrollLeft, setScrollLeft] = useState(0);
-  const [gridWidth, setGridWidth] = useState(0);
+  /**
+   * refs
+   * */
   const gridRef = useRef<HTMLDivElement>(null);
-  const headerRef = useRef<HTMLDivElement>(null);
-  const viewportWidth = (width || gridWidth) - 2; // 2 for border width;
+  const lastSelectedRowIdx = useRef(-1);
 
-  const columnMetrics = useMemo(() => {
-    if (viewportWidth <= 0) return null;
+  /**
+   * states
+   */
+  const [eventBus] = useState(() => new EventBus());
+  const [gridWidth, setGridWidth] = useState(0);
+  const [scrollTop, setScrollTop] = useState(0);
+  const [scrollLeft, setScrollLeft] = useState(0);
+  const [columnWidths, setColumnWidths] = useState<ReadonlyMap<string, number>>(() => new Map());
+
+  /**
+   * computed values
+   */
+  const viewportWidth = (width || gridWidth) - 2; // 2 for border width;
```

Comment:
This will be `-2` on the first render if the width is not specified or when we are using `Autosizer`. I think we should handle these cases within the grid. We can do something like 

```
function DataGrid() {
   const gridRef = useRef<HTMLDivElement>(null);
   const [gridWidth, setGridWidth] = useState(0);
   const viewportWidth = (width || gridWidth) - 2; // 2 for border width;

  useLayoutEffect(() => {
    // Do not calculate the width if width is provided
    if (width) return;
    ...
  }, [width]);
  ...

  return (
    <div
      className="rdg"
      style={{
        width,
        height,
        '--header-row-height': `${headerRowHeight}px`,
        '--filter-row-height': `${headerFiltersHeight}px`,
        '--row-width': `${totalColumnWidth}px`,
        '--row-height': `${rowHeight}px`
      } as React.CSSProperties}
      ref={gridRef}
      onScroll={onScroll}
    >
       {viewportWidth > 0 &&  <Viewport ... />} or // Canvas
  </div>
}
```

This will fix the scrollbar issue with Autosizer and we won't calculate columns until we have a valid width. Thoughts?

Reply:
Feels like a big refactor for very little gain.